common/djangoapps/student/models.py

```diff
@@ -1605,7 +1613,13 @@ def enrollments_for_user_with_overviews_preload(cls, user):  # pylint: disable=i
         The name of this method is long, but was the end result of hashing out a
         number of alternatives, so pylint can stuff it (disable=invalid-name)
         """
+
         enrollments = list(cls.enrollments_for_user(user))
+        if not load_all_courses:
+            courses_count = cls.get_dashboard_course_limit()
+            if courses_count:
+                enrollments = list(cls.enrollments_for_user(user).order_by('-created'))[:courses_count]
```

Comment:
Casting this as a `list` and then doing the slicing will result  in the following:

1. Get all the enrollements from MySQL and put them in memory in a list object.
2. Truncate the list object to our course list limit.

MySQL can just pass you back only the number of items you wanted if you just add the [array slicing to the underlying queryset](https://docs.djangoproject.com/en/1.11/topics/db/queries/#limiting-querysets).

Reply:
In addition to @feanil's points, I believe doing this order by is going to result in a filesort, even with the limit, because there is no explicit `(user_id, -created)` index for it to key off of. Even if individual queries finish quickly, the dashboard is such a frequent operation that it may cause us problems. Can you please double check the query that the Django ORM generates here and run an `explain` on it using the read replica?

Reply:
@ormsbee yes its using filesort, Checkout query explanation in attached image.
<img width="1439" alt="Screen Shot 2019-06-12 at 3 31 23 PM" src="https://user-images.githubusercontent.com/8156837/59344356-5cb33480-8d27-11e9-93ca-5513d6636ccf.png">

Reply:
Then please find a way to make a query for this that doesn't use filesort, or create a migration that adds the appropriate `(user_id, -created)` composite index. Enrollments are in a large table, so you should let @edx/devops know if such a migration is coming so they can advise and we don't bring down the site when adding the index. But that's better than putting this up and then the site suddenly cratering because we suddenly go over our IOPS limit with RDS.

Reply:
@ormsbee how about if we use `sort` instead of order_by ?
```suggestion
 enrollments = list(cls.enrollments_for_user(user))
 (enrollments.sort(key=lambda x: x.created, reverse=True))[:courses_count]
                
```

Reply:
Locally I have checked query time using sort and order by in django debug tool bar. And with sort it takes less time. checkout attached screen shots.
**Using sort:**
<img width="1234" alt="Screen Shot 2019-06-13 at 7 30 50 PM" src="https://user-images.githubusercontent.com/8156837/59441479-15ee3900-8e12-11e9-8dde-5ce0b7ae439c.png">
**Using order_by:**
<img width="1234" alt="Screen Shot 2019-06-13 at 7 16 55 PM" src="https://user-images.githubusercontent.com/8156837/59441535-2e5e5380-8e12-11e9-8ef7-26f4bcf2a1a6.png">

Reply:
How many courses were you testing with those screenshots?

The main issue with instantiating it into a list and then running sort on it is that it means we have to pull all those rows from the database first and then run them through the Django ORM. We have people who are enrolled in thousands of courses. Now it's possible that pulling all those rows out of the database isn't our limiting factor and that it'll have acceptable performance because we'd prune it before we get to more advanced processing. As bad as pulling an unbounded query like this from the database is, it's no worse than what exists today, and it does not introduce a filesort, so it shouldn't increase our operational risk there.

That being said, I would strongly advise that you create a new composite index to make the query we really want run without the filesort. It's a more fundamental fix that enables a very common query pattern, it's less surprising, and it's not that much more work.